File: aat-android/src/main/java/ch/bailu/aat/services/sensor/bluetooth_le/ServiceInterface.java
package ch.bailu.aat.services.sensor.bluetooth_le;

import android.bluetooth.BluetoothGattCharacteristic;

import ch.bailu.aat_lib.gpx.GpxInformation;

public interface ServiceInterface {
    void close();

    void changed(BluetoothGattCharacteristic c);

    boolean isValid();

    boolean discovered(BluetoothGattCharacteristic c, Executer execute);

    void read(BluetoothGattCharacteristic c);

    GpxInformation getInformation(int iid);
}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/sensor/bluetooth_le/ServiceInterface.kt
package ch.bailu.aat.services.sensor.bluetooth_le

import android.bluetooth.BluetoothGattCharacteristic
import ch.bailu.aat_lib.gpx.information.GpxInformation

/**
 * Contract for a BLE GATT sensor service (e.g. Cycling Power, Heart Rate).
 *
 * Lifecycle: [discovered] registers characteristics -> [read] captures static
 * features -> [changed] delivers live measurements -> [getInformation] lets
 * callers query the latest [GpxInformation] by [InfoID].
 */
interface ServiceInterface {
    fun close()
    fun changed(c: BluetoothGattCharacteristic)
    val isValid: Boolean
    fun discovered(c: BluetoothGattCharacteristic, execute: Executor): Boolean
    fun read(c: BluetoothGattCharacteristic)
    fun getInformation(iid: Int): GpxInformation?
}
